pkg/controller/atlasdeployment/search_nodes.go

```diff
@@ -0,0 +1,312 @@
+			s.ctx.Context, s.projectID, s.deployment.GetDeploymentName(), &admin.ApiSearchDeploymentRequest{
+				Specs: s.deployment.Spec.DeploymentSpec.SearchNodesToAtlas(),
+			}).Execute()
+		if err != nil {
+			return s.terminate(workflow.ErrorSearchNodesNotUpsertedInAtlas, err)
+		}
+
+		return s.progress(
+			workflow.SearchNodesUpdating,
+			fmt.Sprintf("search nodes are not ready yet: Atlas state is %q", updateResponse.GetStateName()),
+			"updating search nodes",
+		)
+	}
+	s.ctx.Log.Debug("search nodes in AKO and Atlas are equal")
+
+	// even if both Atlas and AKO are equal, check if existing atlas nodes are potentially are not IDLE (i.e. UPDATING).
+	// In this case we should continue staying in "updating" state.
+	if atlasNodes.GetStateName() != "IDLE" {
+		return s.progress(
+			workflow.SearchNodesUpdating,
+			fmt.Sprintf("search nodes are not ready yet: Atlas state is %q", atlasNodes.GetStateName()),
+			"updating search nodes",
+		)
+	}
+
+	// no changes needed, continue transitioning to pending state, marking search node status as idle.
+	return s.idle()
+}
+
+// delete deletes search nodes in Atlas. It transitions to the following states:
+// - deleting: after search nodes have been deleted in Atlas.
+// - terminated: when an error occurred.
+func (s *searchNodeController) delete() workflow.Result {
+	s.ctx.Log.Debug("deleting search nodes")
+	_, err := s.ctx.SdkClient.AtlasSearchApi.DeleteAtlasSearchDeployment(s.ctx.Context, s.projectID, s.deployment.GetDeploymentName()).Execute()
+	if err != nil {
+		return s.terminate(workflow.ErrorSearchNodesNotDeletedInAtlas, err)
+	}
+
+	return s.progress(
+		workflow.SearchNodesDeleting,
+		"deleting search nodes",
+		"deleting search nodes",
+	)
+}
+
+// progress transitions to the given state in the "SearchNodesReady" status and sets the given fineMsg as its message.
+// further it returns a coarse grained progress state for bubbling the chain.
+func (s *searchNodeController) progress(state workflow.ConditionReason, fineMsg, coarseMsg string) workflow.Result {
+	var (
+		fineProgress   = workflow.InProgress(state, fineMsg)
+		coarseProgress = workflow.InProgress(state, coarseMsg)
+	)
+
+	s.ctx.SetConditionFromResult(status.SearchNodesReadyType, fineProgress)
+	return coarseProgress
+}
+
+// terminate transitions to pending state if an error occurred.
+func (s *searchNodeController) terminate(reason workflow.ConditionReason, err error) workflow.Result {
+	s.ctx.Log.Error(err)
+	result := workflow.Terminate(reason, err.Error())
+	s.ctx.SetConditionFromResult(status.SearchNodesReadyType, result)
+	return result
+}
+
+// unmanage transitions to pending state if no search nodes are managed.
+func (s *searchNodeController) unmanage() workflow.Result {
+	s.ctx.UnsetCondition(status.SearchNodesReadyType)
+	return workflow.OK()
+}
+
+// unmanage transitions to pending state search nodes are ready and idle.
+func (s *searchNodeController) idle() workflow.Result {
```

Comment:
This seems a copy paste of the previous comment, we should explain idle instead here

Reply:
oh indeed, @roothorp do you mind to fix this here?

Reply:
@roothorp please take a look. Add the short explanation of the state like here: https://github.com/mongodb/mongodb-atlas-kubernetes/pull/1392/files/728e0fb5ca88b0f2bfde945572bc452363df46f2#diff-f0151c74d09af63db28986ce11d51e640ffeffd797faf035ea9f2ef390e11541R208-R209